File: packages/core/src/engine/underdog.ts
import type { UnderdogPolicy } from '../config/policy';
import { clampBig, maxBig, minBig } from '../math/fixed';
import { ErrorCode, fail } from './errors';
import { instantShareAfterDepositBps, twabShareBps, type ShareInputs } from './twab';
import type { SideId } from './types';

/**
 * Underdog multiplier (ECONOMICS §7.1–7.2), Q4 fixed point (10_000 = 1.0×).
 */

export const ONE_Q4 = 10_000n;

/** raw_m = clamp(1 + slope × (0.5 − share), 1, cap) in Q4. */
export function rawMultiplierQ4(shareEffBps: bigint, policy: UnderdogPolicy): bigint {
  const slope = BigInt(policy.slope);
  const cap = BigInt(policy.capQ4);
  const raw = ONE_Q4 + slope * (5000n - shareEffBps); // may be < ONE_Q4 when share > 50%
  return clampBig(raw, ONE_Q4, cap);
}

/** warmup_secs = max(duration × warmup_bps / 10_000, warmup_floor_secs). */
export function warmupSecs(durationSecs: bigint, policy: UnderdogPolicy): bigint {
  return maxBig(
    (durationSecs * BigInt(policy.warmupBps)) / 10_000n,
    BigInt(policy.warmupFloorSecs),
  );
}

/** Linear ramp of the boost over the warm-up window. */
export function rampMultiplierQ4(rawQ4: bigint, elapsedSecs: bigint, warmup: bigint): bigint {
  if (elapsedSecs < 0n) fail(ErrorCode.TooEarly, 'elapsed < 0');
  if (warmup === 0n) return rawQ4;
  const boost = rawQ4 - ONE_Q4;
  return ONE_Q4 + (boost * minBig(elapsedSecs, warmup)) / warmup;
}

export interface MultiplierInputs extends ShareInputs {
  side: SideId;
  depositUnits: bigint;
  now: bigint;
  startTs: bigint;
  endTs: bigint;
  policy: UnderdogPolicy;
}

export interface MultiplierResult {
  mQ4: bigint;
  rawQ4: bigint;
  shareInstBps: bigint;
  shareTwabBps: bigint | null;
  shareEffBps: bigint;
  warmup: bigint;
  elapsed: bigint;
}

/**
 * Multiplier for a tranche deposited now. Side aggregates must already be
 * accrued to `now`.
 *
 * share_eff = max(instant share after deposit, TWAB share before deposit);
 * when no time has accrued yet the TWAB is undefined and the instant share is
 * used alone (the warm-up ramp then neutralises it).
 */
export function underdogMultiplier(inp: MultiplierInputs): MultiplierResult {
  const shareInstBps = instantShareAfterDepositBps(inp, inp.side, inp.depositUnits);
  const shareTwab = twabShareBps(inp, inp.side);
  const shareEffBps = shareTwab === null ? shareInstBps : maxBig(shareInstBps, shareTwab);
  const rawQ4 = rawMultiplierQ4(shareEffBps, inp.policy);
  const warmup = warmupSecs(inp.endTs - inp.startTs, inp.policy);
  const elapsed = inp.now - inp.startTs;
  const mQ4 = rampMultiplierQ4(rawQ4, elapsed, warmup);
  return { mQ4, rawQ4, shareInstBps, shareTwabBps: shareTwab, shareEffBps, warmup, elapsed };
}

File: programs/tribe_arena/src/engine/underdog.rs
//! Underdog multiplier — transcription of `packages/core/src/engine/underdog.ts`
//! (ECONOMICS §7.1–7.2), Q4 fixed point.

use crate::engine::math::{mul_u128, ONE_Q4};
use crate::engine::shares::{instant_share_after_deposit_bps, twab_share_bps, Valuation};
use crate::errors::TribeError;
use crate::state::{SideState, UnderdogPolicy};
use anchor_lang::prelude::*;

/// `raw_m = clamp(1 + slope × (0.5 − share), 1, cap)` in Q4.
pub fn raw_multiplier_q4(share_eff_bps: u128, policy: &UnderdogPolicy) -> u128 {
    let slope = policy.slope as i128;
    let cap = policy.cap_q4 as i128;
    let raw = ONE_Q4 as i128 + slope * (5000 - share_eff_bps as i128);
    raw.clamp(ONE_Q4 as i128, cap) as u128
}

/// `warmup = max(duration × warmup_bps / 10_000, warmup_floor_secs)`.
pub fn warmup_secs(duration_secs: i64, policy: &UnderdogPolicy) -> u128 {
    let d = duration_secs.max(0) as u128;
    let pct = d * policy.warmup_bps as u128 / 10_000;
    pct.max(policy.warmup_floor_secs.max(0) as u128)
}

/// Linear ramp of the boost over the warm-up window.
pub fn ramp_multiplier_q4(raw_q4: u128, elapsed_secs: i64, warmup: u128) -> Result<u128> {
    require!(elapsed_secs >= 0, TribeError::TooEarly);
    if warmup == 0 {
        return Ok(raw_q4);
    }
    let boost = raw_q4 - ONE_Q4;
    let e = (elapsed_secs as u128).min(warmup);
    Ok(ONE_Q4 + mul_u128(boost, e)? / warmup)
}

pub struct MultiplierResult {
    pub m_q4: u128,
    pub raw_q4: u128,
    pub share_inst_bps: u128,
    pub share_twab_bps: Option<u128>,
    pub share_eff_bps: u128,
    pub warmup: u128,
    pub elapsed: i64,
}

/// Multiplier for a tranche deposited now (sides already accrued to `now`).
#[allow(clippy::too_many_arguments)]
pub fn underdog_multiplier(
    sides: &[SideState; 2],
    vals: &[Valuation; 2],
    side: usize,
    deposit_units: u64,
    now: i64,
    start_ts: i64,
    end_ts: i64,
    policy: &UnderdogPolicy,
) -> Result<MultiplierResult> {
    let share_inst_bps = instant_share_after_deposit_bps(sides, vals, side, deposit_units)?;
    let share_twab = twab_share_bps(sides, vals, side)?;
    let share_eff_bps = match share_twab {
        None => share_inst_bps,
        Some(t) => share_inst_bps.max(t),
    };
    let raw_q4 = raw_multiplier_q4(share_eff_bps, policy);
    let warmup = warmup_secs(end_ts - start_ts, policy);
    let elapsed = now - start_ts;
    let m_q4 = ramp_multiplier_q4(raw_q4, elapsed, warmup)?;
    Ok(MultiplierResult {
        m_q4,
        raw_q4,
        share_inst_bps,
        share_twab_bps: share_twab,
        share_eff_bps,
        warmup,
        elapsed,
    })
}
